sys/arduino/include/spiport.hpp

```diff
@@ -0,0 +1,231 @@
+ */
+
+#ifndef SPIPORT_H
+#define SPIPORT_H
+
+#include "arduino_board.h"
+#include "byteorder.h"
+#include "periph/spi.h"
+#include "rmutex.h"
+
+/**
+ * @name    Arduino compatible SPI modes
+ * @{
+ */
+#define SPI_MODE0       (0)     /**< CPOL=0, CPHA=0 */
+#define SPI_MODE1       (1)     /**< CPOL=0, CPHA=1 */
+#define SPI_MODE2       (2)     /**< CPOL=1, CPHA=0 */
+#define SPI_MODE3       (3)     /**< CPOL=1, CPHA=1 */
+/** @} */
+
+/**
+ * @name    Arduino compatible SPI frequency selection via clock divider
+ *
+ * This API assumes the library was targeting a 16 MHz Arduino. It will choose
+ * the SPI clock frequency matching the requested frequency best, but never
+ * a frequency greater than the one it would have on Arduinos.
+ * @{
+ */
+#define SPI_CLOCK_DIV2      (0) /**< Best match for 8 MHz: 5 MHz */
+#define SPI_CLOCK_DIV4      (1) /**< Best match for 4 MHz: 1 MHz */
+#define SPI_CLOCK_DIV8      (1) /**< Best match for 2 MHz: 1 MHz */
+#define SPI_CLOCK_DIV16     (1) /**< Best match for 1 MHz: 1 MHz */
+#define SPI_CLOCK_DIV32     (2) /**< Best match for 500 kHz: 400 kHz */
+#define SPI_CLOCK_DIV64     (3) /**< Best match for 250 kHZ: 100 kHz */
+#define SPI_CLOCK_DIV128    (3) /**< Best match for 125 kHz: 100 kHz */
+/** @} */
+
+/**
+ * @name    Arduino compatible bit order values for SPI
+ * @{
+ */
+#define MSBFIRST        (0)     /**< most significat bit first */
+/** @} */
+
+/**
+ * @brief   Arduino SPI configuration interface
+ */
+class SPISettings {
+private:
+    spi_mode_t mode;
+    spi_clk_t clock;
+
+public:
+    /**
+     * @brief   Create a new SPI settings instance
+     *
+     * @param   clock_hz    SPI clock in Hz to use
+     * @param   bitOrder    Has to be `MSBFIST`, for compatibility only
+     * @param   dataMode    SPI mode to use
+     *
+     * The RIOT SPI clock frequency best matching @p clock will be chosen,
+     * but the frequency will never by greater than what is given in @p clock,
+     * unless @p clock_hz is is lower than 100kHz, which is the lowest clock
+     * frequency that will be used.
+     */
+    SPISettings(uint32_t clock_hz, uint8_t bitOrder, uint8_t dataMode);
+
+    /**
+     * @brief   Create a new SPI settings instance with default settings
+     */
+    SPISettings() : SPISettings(4000000, MSBFIRST, SPI_MODE0) { }
+
+    friend class SPIClass;
+};
+
+/**
+ * @brief   Arduino SPI interface
+ *
+ * @warning Wrap all SPI transfers in `SPI.beginTransaction()` and `SPI.endTransaction()`
+ *
+ * The official Arduino SPI-API allows to use SPI transfers without having to
+ * call `SPI.beginTransaction()`, but discourages to do so. The RIOT API does
+ * not provide this feature, instead a call to `SPI.beginTransaction()` is
+ * mandatory. However, most Arduino code already does this and the remaining
+ * code should be fixed anyway.
+ */
+class SPIClass {
+private:
+    spi_t spi_dev;
+    SPISettings settings;
+    bool is_transaction;
+    rmutex_t mut;
```

Comment:
Just for clarification. Do we need this extra mutex to avoid that a thread calls `SPI::transfer` and in turn `spi_transfer_bytes` directly while another thread has already started a transaction with `SPI::beginTransaction` and `SPI::is_transaction` is set therefore?

Reply:
The `rmutex` is protecting `is_transaction` and used to synchronize access to the SPI class instance.